In [4]:

!pip install torch-geometric -q
!pip install rdkit
import torch
from torch.utils.data import IterableDataset
from torch_geometric.datasets import MoleculeNet
import pandas as pd
import itertools
import sys
from google.colab import drive
import os
#I presume you placed the 'Project_HIV' folder within your main drive folder.
gdrive_path='/content/gdrive/MyDrive/Project_HIV'


drive.mount('/content/gdrive', force_remount=True)

os.chdir(gdrive_path)

print(sorted(os.listdir()))


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 kB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 31.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.2/36.2 MB 72.3 MB/s eta 0:00:00
Mounted at /content/gdrive
['.git', '.gitignore', 'HIV.ipynb', 'README.md', 'Refactor_HIV.ipynb', '__pycache__', 'code', 'data_import.py', 'datasets', 'hiv_test.jsonl', 'hiv_train.jsonl', 'hiv_val.jsonl', 'model_with_pe.pth', 'model_without_pe.pth', 'utils.py']


In [5]:
from torch_geometric.datasets import MoleculeNet
class PartialHIVDataset(IterableDataset):
    """
    IterableDataset that loads only a small portion of the HIV dataset.
    """

    def __init__(self, root='/tmp/HIV', max_samples=10):
        """
        Args:
            root: Root directory for dataset storage
            max_samples: Maximum number of samples to load
        """
        self.root = root
        self.max_samples = max_samples
        self._dataset = None

    def _lazy_load_dataset(self):
        """Lazy load the dataset only when iteration begins"""
        if self._dataset is None:
            print(f"Initializing HIV dataset (will only load {self.max_samples} samples)...")
            # Import here to ensure it's available

            self._dataset = MoleculeNet(root=self.root, name='HIV')
            print(f"Dataset ready! Total size: {len(self._dataset)} molecules")
            print(f"But we'll only load {self.max_samples} of them.\n")

    def parse_molecules(self):
        """
        Parse molecules from the dataset, stopping after max_samples.
        """
        self._lazy_load_dataset()

        for i in range(min(self.max_samples, len(self._dataset))):
            data = self._dataset[i]

            # Extract molecule information
            mol_info = {
                'source': f"molecule_{i}",
                'target': data.y.item(),    # HIV activity label
                'num_atoms': data.num_nodes,
                'num_bonds': data.num_edges // 2,
                'smiles': getattr(data, 'smiles', 'N/A')
            }

            yield mol_info

    def __iter__(self):
        """Iterator with worker support"""
        iterator = self.parse_molecules()
        worker_info = torch.utils.data.get_worker_info()

        if worker_info is not None:
            worker_total_num = worker_info.num_workers
            worker_id = worker_info.id
            return itertools.islice(iterator, worker_id, None, worker_total_num)

        return iterator


# Your code
batch_list = []
batch_size = 40

# Create dataset
dataset = PartialHIVDataset(max_samples=batch_size)
iterator = iter(dataset)
full_batch_data = list(dataset)

df = pd.DataFrame(full_batch_data)

Initializing HIV dataset (will only load 40 samples)...


Processing...
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'O=C1O[Al]23(OC1=O)(OC(=O)C(=O)O2)OC(=O)C(=O)O3' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'Cc1ccc([B-2]2(c3ccc(C)cc3)=NCCO2)cc1' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'Oc1ccc(C2Oc3cc(O)cc4c3C(=[O+][AlH3-3]35([O+]=C6c7c(cc(O)cc7[OH+]3)OC(c3ccc(O)cc3O)C6O)([O+]=C3c6c(cc(O)cc6[OH+]5)OC(c5ccc(O)cc5O)C3O)[OH+]4)C2O)c(O)c1' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'CC1=C2

Dataset ready! Total size: 41120 molecules
But we'll only load 40 of them.



In [6]:
import codecs
import numpy as np
import torch.nn as nn
from transformers import AutoTokenizer

smiles_list = []
batch_size = 40

for _ in range(batch_size):
  mol_info = next(iterator)
  smiles_list.append(mol_info['smiles'])

tokenizer = AutoTokenizer.from_pretrained("seyonec/ChemBERTa-zinc-base-v1")

tokenized_lengths = []
for smiles in smiles_list:
    tokens = tokenizer(smiles, padding=False, truncation=False)
    tokenized_lengths.append(len(tokens['input_ids']))

max_length = max(tokenized_lengths)
print(f"Maximum sequence length in data: {max_length}")

encoded = tokenizer(
    smiles_list,
    padding='max_length',
    truncation=True,
    max_length=max_length,
    return_tensors='pt'
)

embedding = nn.Embedding(
    num_embeddings=len(tokenizer),
    embedding_dim=512,
    padding_idx=tokenizer.pad_token_id
)

embeddings = embedding(encoded['input_ids'])
print(f"Embeddings shape: {embeddings.shape}")  # (batch_size, max_length, 512)



tokenizer_config.json:   0%|          | 0.00/166 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/501 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

Maximum sequence length in data: 61
Embeddings shape: torch.Size([40, 61, 512])


In [7]:
from scipy.sparse.linalg import eigsh
from rdkit import Chem
def smiles_to_adjacency_matrix(smiles):
    """Convert SMILES to adjacency matrix."""
    mol = Chem.MolFromSmiles(smiles)
    if mol is None:
        return None

    n_atoms = mol.GetNumAtoms()
    adj_matrix = np.zeros((n_atoms, n_atoms))

    for bond in mol.GetBonds():
        i = bond.GetBeginAtomIdx()
        j = bond.GetEndAtomIdx()
        # Undirected graph
        adj_matrix[i, j] = 1
        adj_matrix[j, i] = 1

    return adj_matrix
#k is the number of eigenvectors
def compute_graph_positional_encoding(adj_matrix, k=30):
    """
    Compute graph positional encodings from eigenvectors of the
    symmetrically normalized graph Laplacian.
    """
    n = adj_matrix.shape[0]

    # Compute degree matrix
    degree = np.sum(adj_matrix, axis=1)
    degree[degree == 0] = 1

    # D^(-1/2)
    d_inv_sqrt = np.diag(1.0 / np.sqrt(degree))

    # Symmetrically normalized Laplacian
    identity = np.eye(n)
    normalized_adj = d_inv_sqrt @ adj_matrix @ d_inv_sqrt
    laplacian = identity - normalized_adj

    # Always use eigh for small matrices or when k >= n-1
    if n <= k + 1:  # Changed condition
        eigenvalues, eigenvectors = np.linalg.eigh(laplacian)
        if n < k:
            # Pad with zeros
            padded_eigenvectors = np.zeros((n, k))
            padded_eigenvectors[:, :n] = eigenvectors
            return padded_eigenvectors
        else:
            return eigenvectors[:, :k]
    else:
        # Use sparse solver only when k < n-1
        eigenvalues, eigenvectors = eigsh(laplacian, k=k, which='SM')
        return eigenvectors

In [8]:
# Process all SMILES and compute their graph PEs
graph_pes_list = []
max_nodes = 0

for smiles in smiles_list:
    adj_matrix = smiles_to_adjacency_matrix(smiles)
    if adj_matrix is not None:
        pe = compute_graph_positional_encoding(adj_matrix, k=30)
        graph_pes_list.append(pe)
        max_nodes = max(max_nodes, pe.shape[0])
    else:
        graph_pes_list.append(None)

print(f"Maximum number of atoms in dataset: {max_nodes}")


Maximum number of atoms in dataset: 42


In [9]:
import torch
import numpy as np
from rdkit import Chem
import re



def generate_zero_pe(n_vectors, dim=30):
    """Generate n_vectors zero vectors of dimension dim for non-atom characters."""
    return np.zeros((n_vectors, dim))

def parse_token_components(token):
    """
    Parse a token to identify atoms and characters.
    Returns: (atom_indices, n_characters)
    """
    # Common atom patterns in SMILES
    atom_pattern = r'(Cl|Br|Si|Mg|Ca|Fe|Al|Na|Li|[BCNOFPSKHIV])'

    # Find all atoms in the token
    atoms = re.findall(atom_pattern, token)

    # Count non-atom characters
    # Remove atoms from token to count remaining characters
    remaining = token
    for atom in atoms:
        remaining = remaining.replace(atom, '', 1)

    # Count actual characters (digits, +, -, =, #, etc.)
    n_characters = len([c for c in remaining])

    return atoms, n_characters

def align_graph_pe_to_tokens(smiles, graph_pe, tokenizer, max_length, random_seed=None):
    """
    Align graph PE to tokens, handling pure characters, atoms, and mixed tokens.
    """
    if random_seed is not None:
        np.random.seed(random_seed)

    # Determine embedding dimension
    if graph_pe is not None:
        embedding_dim = graph_pe.shape[1]
    else:
        embedding_dim = 30  # Default fallback

    mol = Chem.MolFromSmiles(smiles)
    if mol is None or graph_pe is None:
        return torch.zeros(max_length, embedding_dim)

    # Get tokens
    encoding = tokenizer(smiles, padding='max_length', max_length=max_length)
    tokens = tokenizer.convert_ids_to_tokens(encoding['input_ids'])
    token_pe = torch.zeros(len(tokens), embedding_dim)

    # Build atom mapping from SMILES
    atom_symbols = [atom.GetSymbol() for atom in mol.GetAtoms()]
    atom_count = {symbol: 0 for symbol in set(atom_symbols)}
    atom_to_idx = {}

    for idx, symbol in enumerate(atom_symbols):
        atom_to_idx[(symbol, atom_count[symbol])] = idx
        atom_count[symbol] += 1

    # Reset atom count for tracking
    current_atom_count = {symbol: 0 for symbol in atom_count}

    # Process each token
    for i, token in enumerate(tokens):
        if token in ['<s>', '</s>', '<pad>']:
            continue

        # Parse token components
        atoms_in_token, n_characters = parse_token_components(token)

        if len(atoms_in_token) == 0 and n_characters > 0:
            # Pure character token - use zero PE
            random_pes = generate_zero_pe(n_characters, embedding_dim)
            token_pe[i] = torch.tensor(random_pes.sum(axis=0))

        elif len(atoms_in_token) > 0 and n_characters == 0:
            # Pure atom token(s) - use graph PE
            atom_pes = []
            for atom_symbol in atoms_in_token:
                if atom_symbol in current_atom_count:
                    atom_key = (atom_symbol, current_atom_count[atom_symbol])
                    if atom_key in atom_to_idx:
                        atom_idx = atom_to_idx[atom_key]
                        atom_pes.append(graph_pe[atom_idx])
                        current_atom_count[atom_symbol] += 1

            if atom_pes:
                # sum the PEs of all atoms in this token
                token_pe[i] = torch.tensor(np.sum(atom_pes, axis=0))

        elif len(atoms_in_token) > 0 and n_characters > 0:
            # Mixed token - combine atom PE and character PE
            atom_pes = []
            for atom_symbol in atoms_in_token:
                if atom_symbol in current_atom_count:
                    atom_key = (atom_symbol, current_atom_count[atom_symbol])
                    if atom_key in atom_to_idx:
                        atom_idx = atom_to_idx[atom_key]
                        atom_pes.append(graph_pe[atom_idx])
                        current_atom_count[atom_symbol] += 1

            # Get character PEs (zeros as you intended)
            random_pes = generate_zero_pe(n_characters, embedding_dim)

            # Combine: sum of atoms + sum of characters
            combined_pe = np.zeros(embedding_dim)
            if atom_pes:
                combined_pe += np.sum(atom_pes, axis=0)
            combined_pe += random_pes.sum(axis=0)

            token_pe[i] = torch.tensor(combined_pe)

    # Now pad or truncate to max_length
    if len(tokens) < max_length:
        # Pad with zeros
        padded_token_pe = torch.zeros(max_length, embedding_dim)
        padded_token_pe[:len(tokens)] = token_pe
        return padded_token_pe
    else:
        # Truncate
        return token_pe[:max_length]



In [10]:
token_pes_list = []

for smiles, graph_pe in zip(smiles_list, graph_pes_list):
    # Align graph PE to tokens using the provided function
    token_pe = align_graph_pe_to_tokens(
        smiles=smiles,
        graph_pe=graph_pe,
        tokenizer=tokenizer,
        max_length=max_length,
        random_seed=42
    )
    token_pes_list.append(token_pe)

# Stack into batch tensor
token_pes_batch = torch.stack(token_pes_list)

In [11]:
class EmbeddingWithGraphPE(nn.Module):
    def __init__(self, embed_dim=512, pe_dim=30):
        super().__init__()
        # One-layer projection with GeLU from graph PE to embedding dimension
        # Using standard Laplacian, so pe_dim is just k
        self.pe_projection = nn.Sequential(
            nn.Linear(pe_dim, embed_dim),
            nn.GELU()
        )

    def forward(self, embeddings, token_pes):
        # token_pes shape: [batch_size, seq_len, pe_dim]
        # embeddings shape: [batch_size, seq_len, embed_dim]

        # Project token PEs to embedding dimension with GeLU
        projected_pes = self.pe_projection(token_pes)

        # Add to token embeddings
        enhanced_embeddings = embeddings + projected_pes

        return enhanced_embeddings

# Usage
model = EmbeddingWithGraphPE(embed_dim=512, pe_dim=30)
# graph_pes_batch shape: [40, 61, 30] (30 eigenvectors from standard Laplacian)
# embeddings shape: [40, 61, 512]
enhanced_embeddings = model(embeddings, token_pes_batch)
# Output shape: [40, 61, 512]

In [12]:
enhanced_embeddings.shape

torch.Size([40, 61, 512])

In [13]:
import json
from data_import import check_class_balance, stratified_train_val_test_split, convert_to_litgpt_format, save_to_jsonl

# Create dataset
batch_size = 41120
dataset = PartialHIVDataset(max_samples=batch_size)

# Convert to DataFrame
full_batch_data = list(dataset)
df = pd.DataFrame(full_batch_data)

# Check balance in the original DataFrame
original_balance = check_class_balance(df, "Original Dataset")

# Convert to LitGPT format for different task types
classification_data = convert_to_litgpt_format(df, task_type="classification")

# Check balance in the converted data
converted_balance = check_class_balance(classification_data, "Converted Dataset")

# Create stratified train/validation/test split (60/20/20)
train_data, val_data, test_data = stratified_train_val_test_split(
    classification_data,
    train_ratio=0.6,
    val_ratio=0.2,
    test_ratio=0.2
)

# Check balance in all sets
train_balance = check_class_balance(train_data, "Training Set")
val_balance = check_class_balance(val_data, "Validation Set")
test_balance = check_class_balance(test_data, "Test Set")

# Save to JSONL files
save_to_jsonl(train_data, "hiv_train.jsonl")
save_to_jsonl(val_data, "hiv_val.jsonl")
save_to_jsonl(test_data, "hiv_test.jsonl")

print(f"\nDataset sizes:")
print(f"Train: {len(train_data)}")
print(f"Val: {len(val_data)}")
print(f"Test: {len(test_data)}")

Initializing HIV dataset (will only load 41120 samples)...
Dataset ready! Total size: 41120 molecules
But we'll only load 41120 of them.


Original Dataset Class Balance:
Total samples: 41120
Class 0.0: 39677 samples (96.49%)
Class 1.0: 1443 samples (3.51%)

Converted Dataset Class Balance:
Total samples: 41120
Class 0: 39677 samples (96.49%)
Class 1: 1443 samples (3.51%)

Training Set Class Balance:
Total samples: 24671
Class 0: 23806 samples (96.49%)
Class 1: 865 samples (3.51%)

Validation Set Class Balance:
Total samples: 8224
Class 0: 7935 samples (96.49%)
Class 1: 289 samples (3.51%)

Test Set Class Balance:
Total samples: 8225
Class 0: 7936 samples (96.49%)
Class 1: 289 samples (3.51%)
Saved 24671 entries to hiv_train.jsonl
Saved 8224 entries to hiv_val.jsonl
Saved 8225 entries to hiv_test.jsonl

Dataset sizes:
Train: 24671
Val: 8224
Test: 8225


In [14]:
import json
import torch
from utils import calculate_max_lengths
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer

def load_jsonl(filepath):
    """Load data from JSONL file"""
    data = []
    with open(filepath, 'r') as f:
        for line in f:
            data.append(json.loads(line))
    return data

class HIVDataset(Dataset):
    def __init__(self, jsonl_filepath, tokenizer, max_full_length):
        self.data = load_jsonl(jsonl_filepath)
        self.tokenizer = tokenizer
        self.max_full_length = max_full_length

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        item = self.data[idx]

        # Extract SMILES
        smiles = item['input'].replace("SMILES: ", "")

        # Create full prompt
        full_prompt = f"{item['instruction']}\n{item['input']}\nAnswer:"
        # Compute graph PE with actual SMILES length
        smiles_tokens = self.tokenizer(smiles, add_special_tokens=False)
        actual_smiles_length = len(smiles_tokens['input_ids'])
        # Get prompt length for label masking
        prompt_encoding = self.tokenizer(full_prompt, add_special_tokens=True, return_tensors='pt')
        prompt_length = prompt_encoding['input_ids'].shape[1]

        # Full sequence for training
        target_text = f"{full_prompt} {item['output']}"
        full_encoding = self.tokenizer(
            target_text,
            padding='max_length',
            max_length=self.max_full_length,
            truncation=True,
            return_tensors='pt'
        )

        # Compute graph PE for SMILES only
        adj_matrix = smiles_to_adjacency_matrix(smiles)
        if adj_matrix is not None:
            graph_pe = compute_graph_positional_encoding(adj_matrix, k=30)

            smiles_token_pe = align_graph_pe_to_tokens(
                smiles, graph_pe, self.tokenizer, actual_smiles_length, random_seed=42
            )
        else:
            smiles_token_pe = torch.zeros(actual_smiles_length, 30)

        # Create full-length PE tensor
        full_token_pe = torch.zeros(self.max_full_length, 30)

        # Find where SMILES tokens are and place PEs there
        instruction_text = f"{item['instruction']}\nSMILES: "
        instruction_tokens = self.tokenizer(instruction_text, add_special_tokens=True)
        smiles_start_idx = len(instruction_tokens['input_ids'])

        # Copy SMILES PEs to correct position
        pe_end_idx = min(smiles_start_idx + actual_smiles_length, self.max_full_length)
        pe_length = min(smiles_token_pe.shape[0], pe_end_idx - smiles_start_idx)
        full_token_pe[smiles_start_idx:smiles_start_idx + pe_length] = smiles_token_pe[:pe_length]
        # Debug: Check final PE tensor

        # Create labels with prompt masked
        labels = full_encoding['input_ids'][0].clone()
        labels[:prompt_length] = -100  # Mask prompt tokens
        # Also mask padding tokens
        attention_mask = full_encoding['attention_mask'][0]
        labels[attention_mask == 0] = -100
        return {
            'input_ids': full_encoding['input_ids'][0],
            'graph_pes': full_token_pe,
            'labels': labels,
        }

# First calculate max lengths
max_full_length = calculate_max_lengths('hiv_train.jsonl')

def create_dataloaders(train_path='hiv_train.jsonl',
                      val_path='hiv_val.jsonl',
                      test_path='hiv_test.jsonl',  # Add test path
                      batch_size=8,
                      tokenizer=None,
                      max_full_length=max_full_length):
    if tokenizer is None:
        tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-160m")
        if tokenizer.pad_token is None:
            tokenizer.pad_token = tokenizer.eos_token

    train_dataset = HIVDataset(train_path, tokenizer, max_full_length)
    val_dataset = HIVDataset(val_path, tokenizer, max_full_length)
    test_dataset = HIVDataset(test_path, tokenizer, max_full_length)  # Add test dataset

    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)  # Add test loader

    return train_loader, val_loader, test_loader, tokenizer  # Return test_loader too

# Updated usage
train_loader, val_loader, test_loader, tokenizer = create_dataloaders(batch_size=8)

# Test one batch from each loader
print("Train batch:")
sample_batch = next(iter(train_loader))
print(f"Input IDs shape: {sample_batch['input_ids'].shape}")
print(f"Graph PEs shape: {sample_batch['graph_pes'].shape}")
print(f"Labels shape: {sample_batch['labels'].shape}")

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

Train batch:
Input IDs shape: torch.Size([8, 498])
Graph PEs shape: torch.Size([8, 498, 30])
Labels shape: torch.Size([8, 498])


In [15]:
# Look at first example in batch
i = 0
print(f"Input text (decoded): {tokenizer.decode(sample_batch['input_ids'][i])}")
print(f"\nNon-zero PE positions: {torch.nonzero(sample_batch['graph_pes'][i].sum(dim=1)).squeeze().tolist()}")
print(f"\nLabel tokens that aren't -100: {sample_batch['labels'][i][sample_batch['labels'][i] != -100].tolist()}")

Input text (decoded): Classify the following molecule based on its HIV activity. Respond with '1' if the molecule shows HIV activity, or '0' if it does not.
SMILES: CC12CCCCC1=[N+]([O-])C(C)(C)N2O
Answer: 0<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><



---

## Implementation, Results, Comments and Challenges

### 1. Design Decisions and Simplifications

#### Handling Non-Atom Characters
- Set positional encodings (PEs) for non-atom characters to zero
- Preserves structural information from atoms while avoiding noise from SMILES syntax

#### Addressing Class Imbalance
Dataset: 96.5% negative vs 3.5% positive HIV activity

Experimented with three loss functions:
| Loss Function | Result |
|---------------|---------|
| Focal loss | Unstable |
| Dice loss | Produced inf values |
| **Weighted cross-entropy** | **Most stable (selected)** |

#### Training Configuration
- Learning rate warm-up following paper methodology
- Stratified train/val/test split preserving original distribution
- Early stopping at 250 batches for rapid experimentation

### 2. Alignment Function Architecture

Token-to-PE mapping handles three cases:

```
Pure character tokens    →  Zero PE
Pure atom tokens        →  Graph PE from atom
Mixed tokens           →  Sum(atom PEs) + Sum(zero PEs)
```

**Future consideration:** Functional group-level patterns could be enforced, though this risks over-engineering features the network should learn independently.

### 3. Results Analysis

#### Performance Summary

| Model | Class 0 Acc | Class 1 Acc | Test Loss | F1 Score |
|-------|-------------|-------------|-----------|----------|
| **With PE** | 99.99% | 100% | 0.0066 | 0.9983 |
| **Without PE** | 100% | 100% | 0.0063 | 1.0000 |

#### Key Findings

**1. Convergence Speed**
- Both models reach >99% accuracy within 200-250 batches
- Loss drops to <0.01 in 200 iterations
- Suggests task complexity may be insufficient for architecture. This was coroborated when I increased the L2 regularization weight decay to 0.5 to see if there was a degradation in performance. I hypothesized that a higher weight decay should prevent it from learning a complex pattern but the performance remained near perfect. This seems to indicate the task is too easy rather than a case of overfitting. I also checked the activations and they didnt seem to be abnormal:

Layer 0: mean=-0.000, std=0.003, % dead neurons=0.000
Layer 1: mean=0.014, std=1.774, % dead neurons=0.000
Layer 2: mean=1.604, std=3.106, % dead neurons=0.000
etc...

**2. PE Impact**
- No measurable improvement from positional encodings

**3. Possible Explanations**
- HIV activity may depend on simple functional group presence/absence
- SMILES sequential representation already encodes connectivity

#### Validity Concerns

Despite suspicious perfect accuracy on imbalanced data:
- Strong performance on unseen test set contradicts overfitting
- Verified no data leakage between train/val/test splits

# Technical Challenges from HIV Molecular Property Prediction Task

## 1. Graph PE to Token Alignment
**Challenge**: Tokenizer splits molecules unpredictably (e.g., 'Cl' → ['C', 'l'] or ['Cl'])
**Solution**: Built regex parser to identify atoms within tokens:
```python
atom_pattern = r'(Cl|Br|Si|Mg|Ca|Fe|Al|Na|Li|[BCNOFPSKHIV])'
# Three cases: pure chars → zero PE, pure atoms → graph PE, mixed → sum both
```

## 2. Extreme Class Imbalance (96.5% vs 3.5%)
**Challenge**: Models predicted all negatives for easy 96.5% accuracy
**Solution**:
- Tested focal loss (unstable), dice loss (inf values), settled on weighted CE [0.035, 0.965]
- Stratified sampling + per-class metrics tracking

## 3. Suspicious Perfect Performance
**Challenge**: >99% accuracy in 200 batches seemed unrealistic
**Solution**: Diagnostic testing revealed genuine simplicity:
- L2=0.5 → no degradation (not overfitting)
- No dead neurons in activation analysis
- Verified no data leakage via stratified splits
- Consistent performance on held-out test set

## 4. GPU Memory Management
**Challenge**: Sequential model training caused OOM errors
**Solution**:
```python
del model_with_pe; torch.cuda.empty_cache(); gc.collect()
```
Save checkpoints → reload for evaluation

## 5. Silent PE Alignment Failures
**Challenge**: Some molecules produced all-zero encodings without errors
**Solution**: Added debugging + fallback behavior:
```python
print(f"Non-zero PE positions: {torch.nonzero(graph_pes.sum(dim=1))}")
```

## 6. Eigendecomposition Instability
**Challenge**: Sparse eigensolver failed for tiny molecules (1-2 atoms)
**Solution**: Conditional logic based on matrix size:
```python
if n <= k + 1:  # Dense solver for small molecules
    eigenvectors = np.linalg.eigh(laplacian)
    if n < k: pad_with_zeros()

<div align="center">
<img src="https://drive.google.com/uc?id=1lW4OBGNmaqacNgSU03aK7ym0n9jxIicG" alt="Results visualization" width="900"/>
</div>

In [16]:


import matplotlib.pyplot as plt

import torch
import torch.nn.functional as F
import torch.nn as nn
from transformers import AutoModelForCausalLM, AutoTokenizer, get_linear_schedule_with_warmup
from peft import LoraConfig, get_peft_model, TaskType
from typing import Optional

class HIVPEModule(nn.Module):
    def __init__(self, pe_dim=30, embed_dim=768):
        super().__init__()
        self.pe_projection = nn.Linear(pe_dim, embed_dim)
        self.activation = nn.GELU()
        self.dropout = nn.Dropout(0.2)
        nn.init.xavier_uniform_(self.pe_projection.weight, gain=0.3)
        nn.init.zeros_(self.pe_projection.bias)
        self.scale = 1

    def forward(self, embeddings, graph_pes):
        if graph_pes is not None:
            graph_pes = graph_pes.to(embeddings.dtype)
            projected_pes = self.pe_projection(graph_pes)
            projected_pes = self.activation(projected_pes)
            projected_pes = self.dropout(projected_pes)
            return embeddings + self.scale * projected_pes
        return embeddings

class PythiaWithPE(nn.Module):
    def __init__(self, base_model, pe_dim=30,use_pe=True):
        super().__init__()
        self.base_model = base_model
        self.embed_dim = base_model.gpt_neox.embed_in.embedding_dim
        self.use_pe = use_pe
        if self.use_pe:
          self.pe_module = HIVPEModule(pe_dim=pe_dim, embed_dim=self.embed_dim)

    def forward(self, input_ids=None, graph_pes=None, attention_mask=None, labels=None, **kwargs):
        # Store for tracker access
        self.labels = labels
        self.batch = input_ids

        inputs_embeds = self.base_model.gpt_neox.embed_in(input_ids)

        if self.use_pe and graph_pes is not None:
            inputs_embeds = self.pe_module(inputs_embeds, graph_pes)

        kwargs.pop('input_ids', None)

        outputs = self.base_model(
            inputs_embeds=inputs_embeds,
            attention_mask=attention_mask,
            labels=labels,
            **kwargs
        )

        self.output = outputs.logits

        # Calculate loss here instead of in training loop
        if labels is not None:
            self.loss_seg = self._calculate_weighted_loss()

        return outputs

    def _calculate_weighted_loss(self):
        """Add this new method to calculate loss"""
        token_0 = 470  # " 0"
        token_1 = 337  # " 1"

        mask = ((self.labels == token_0) | (self.labels == token_1)) & (self.labels != -100)

        if not mask.any():
            return torch.tensor(0.0, device=self.labels.device)

        logits_flat = self.output.view(-1, self.output.size(-1))
        labels_flat = self.labels.view(-1)
        mask_flat = mask.view(-1)

        relevant_logits = logits_flat[mask_flat]
        relevant_labels = labels_flat[mask_flat]

        ce_loss = F.cross_entropy(relevant_logits, relevant_labels, reduction='none')
        sample_weights = torch.where(
            relevant_labels == token_1,
            torch.tensor(0.965, device=ce_loss.device),
            torch.tensor(0.035, device=ce_loss.device)
        )

        return (ce_loss * sample_weights).mean()

    def generate(self, input_ids, graph_pes=None, **kwargs):
        inputs_embeds = self.base_model.gpt_neox.embed_in(input_ids)
        if self.use_pe and graph_pes is not None:
            inputs_embeds = self.pe_module(inputs_embeds, graph_pes)

        return self.base_model.generate(
            inputs_embeds=inputs_embeds,
            **kwargs
        )
    def get_current_losses(self):
        """Entry point for the tracker to grab the loss"""
        return {"loss_seg": float(self.loss_seg) if hasattr(self, 'loss_seg') else 0.0}

    def get_output(self):
        """Needed by the tracker to get access to the outputs"""
        return self.output if hasattr(self, 'output') else None

    def get_labels(self):
        """Needed by the tracker to access ground truth labels"""
        return self.labels if hasattr(self, 'labels') else None

    def get_batch(self):
        """Get current batch"""
        return self.batch if hasattr(self, 'batch') else None

    def backward(self):
        """Backward pass matching KPConv style"""
        if hasattr(self, 'loss_seg') and self.loss_seg is not None:
            self.loss_seg.backward()

def setup_model_with_pe_and_lora(model_name="EleutherAI/pythia-160m", pe_dim=30,use_pe=True):
    print(f"Loading {model_name}...")
    print(f"Using positional encodings: {use_pe}")
    base_model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype=torch.float32,
        device_map="auto"
    )

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    model = PythiaWithPE(base_model, pe_dim=pe_dim, use_pe=use_pe)

    lora_config = LoraConfig(
        task_type=TaskType.CAUSAL_LM,
        r=4,
        lora_alpha=8,
        lora_dropout=0.3,
        target_modules=["query_key_value", "dense_h_to_4h", "dense_4h_to_h"],
        init_lora_weights="gaussian"
    )

    model.base_model = get_peft_model(model.base_model, lora_config)

    total_params = sum(p.numel() for p in model.parameters())
    trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print(f"Total parameters: {total_params:,}")
    print(f"Trainable parameters: {trainable_params:,}")
    print(f"Trainable %: {100 * trainable_params / total_params:.2f}%")

    return model, tokenizer



# Updated usage code
if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"

    print("Training WITH positional encodings:")
    model_with_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=True)
    model_with_pe = model_with_pe.to(device)

    print("\nTraining WITHOUT positional encodings:")
    model_without_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=False)
    model_without_pe = model_without_pe.to(device)

    print(f"\nModels setup complete and moved to {device}!")
    print("You can now train both models and compare their performance.")

Training WITH positional encodings:
Loading EleutherAI/pythia-160m...
Using positional encodings: True
Total parameters: 162,862,848
Trainable parameters: 539,904
Trainable %: 0.33%

Training WITHOUT positional encodings:
Loading EleutherAI/pythia-160m...
Using positional encodings: False
Total parameters: 162,839,040
Trainable parameters: 516,096
Trainable %: 0.32%

Models setup complete and moved to cuda!
You can now train both models and compare their performance.


In [25]:
import gc
import torch
from tqdm.auto import tqdm
import time
from transformers import get_linear_schedule_with_warmup

class HIVTracker:
    """Simple tracker for HIV classification metrics"""

    def __init__(self):
        self.reset()

    def reset(self, stage="train"):
        """Reset metrics for new epoch"""
        self.stage = stage
        self.losses = []
        self.class_0_correct = 0
        self.class_1_correct = 0
        self.class_0_total = 0
        self.class_1_total = 0

    def track(self, model):
        """Track metrics from model outputs"""
        # Track loss
        losses = model.get_current_losses()
        self.losses.append(losses['loss_seg'])

        # Track accuracy
        with torch.no_grad():
            output = model.get_output()
            labels = model.get_labels()

            if output is None or labels is None:
                return

            token_0 = 470  # " 0"
            token_1 = 337  # " 1"

            mask = ((labels == token_0) | (labels == token_1)) & (labels != -100)

            if mask.any():
                logits_flat = output.view(-1, output.size(-1))
                labels_flat = labels.view(-1)
                mask_flat = mask.view(-1)

                relevant_logits = logits_flat[mask_flat]
                relevant_labels = labels_flat[mask_flat]

                preds = relevant_logits.argmax(dim=-1)

                class_0_mask = relevant_labels == token_0
                class_1_mask = relevant_labels == token_1

                if class_0_mask.any():
                    self.class_0_correct += (preds[class_0_mask] == relevant_labels[class_0_mask]).sum().item()
                    self.class_0_total += class_0_mask.sum().item()

                if class_1_mask.any():
                    self.class_1_correct += (preds[class_1_mask] == relevant_labels[class_1_mask]).sum().item()
                    self.class_1_total += class_1_mask.sum().item()

    def get_metrics(self):
        """Get current metrics as dictionary"""
        avg_loss = sum(self.losses) / len(self.losses) if self.losses else 0
        class_0_acc = self.class_0_correct / self.class_0_total if self.class_0_total > 0 else 0
        class_1_acc = self.class_1_correct / self.class_1_total if self.class_1_total > 0 else 0

        return {
            f"{self.stage}_loss": avg_loss,
            f"{self.stage}_class0_acc": class_0_acc,
            f"{self.stage}_class1_acc": class_1_acc
        }

    def publish(self, epoch):
        """Print epoch summary"""
        metrics = self.get_metrics()
        print(f"\n{self.stage.upper()} - Epoch {epoch}")
        print(f"  Loss: {metrics[f'{self.stage}_loss']:.4f}")
        print(f"  Class 0 Acc: {metrics[f'{self.stage}_class0_acc']:.2%}")
        print(f"  Class 1 Acc: {metrics[f'{self.stage}_class1_acc']:.2%}")

        # Return metrics for later use
        return {
            'loss': metrics[f'{self.stage}_loss'],
            'class_0_acc': metrics[f'{self.stage}_class0_acc'],
            'class_1_acc': metrics[f'{self.stage}_class1_acc']
        }


class Trainer:
    """Enhanced trainer with full lifecycle management"""

    def __init__(self, model, train_loader, val_loader, test_loader=None,
                 num_epochs=3, learning_rate=1e-5, device=torch.device('cuda'),
                 max_batches=None):
        self.num_epochs = num_epochs
        self._model = model.to(device)
        self.train_loader = train_loader
        self.val_loader = val_loader
        self.test_loader = test_loader  # Add test loader
        self.device = device
        self.learning_rate = learning_rate
        self.tracker = HIVTracker()
        self.max_batches = max_batches

        # Store training history
        self.history = {
            'train_loss': [],
            'train_class0_acc': [],
            'train_class1_acc': [],
            'val_loss': [],
            'val_class0_acc': [],
            'val_class1_acc': []
        }

    def fit(self):
        """Main training loop that returns final test results"""
        # Setup optimizer
        self.optimizer = torch.optim.AdamW(
            self._model.parameters(),
            lr=self.learning_rate,
            weight_decay=0.5
        )

        # Setup scheduler
        num_training_steps = len(self.train_loader) * self.num_epochs
        if self.max_batches:
            num_training_steps = min(num_training_steps, self.max_batches * self.num_epochs)

        self.scheduler = get_linear_schedule_with_warmup(
            self.optimizer,
            num_warmup_steps=100,
            num_training_steps=num_training_steps
        )

        # Training epochs
        for epoch in range(self.num_epochs):
            print(f"\n=========== EPOCH {epoch + 1} ===========")
            time.sleep(0.5)

            # Train
            self.train_epoch()
            train_metrics = self.tracker.publish(epoch + 1)

            # Update history
            self.history['train_loss'].append(train_metrics['loss'])
            self.history['train_class0_acc'].append(train_metrics['class_0_acc'])
            self.history['train_class1_acc'].append(train_metrics['class_1_acc'])

            # Validate
            self.val_epoch()
            val_metrics = self.tracker.publish(epoch + 1)

            # Update history
            self.history['val_loss'].append(val_metrics['loss'])
            self.history['val_class0_acc'].append(val_metrics['class_0_acc'])
            self.history['val_class1_acc'].append(val_metrics['class_1_acc'])

        # Final test evaluation (limited to 50 batches)
        if self.test_loader is not None:
            print("\n" + "="*50)
            print("FINAL TEST SET EVALUATION")
            print("="*50)
            test_results = self.evaluate(self.test_loader, "test",max_batches=50)
            return test_results
        else:
            print("\nNo test loader provided, returning validation results")
            return val_metrics

    def train_epoch(self):
      """Training epoch with mixed precision"""
      self._model.train()
      self.tracker.reset("train")

      # Initialize gradient scaler for mixed precision
      scaler = torch.cuda.amp.GradScaler()

      with tqdm(self.train_loader, desc="Training") as tq_train_loader:
          for i, data in enumerate(tq_train_loader):
              if self.max_batches and i >= self.max_batches:
                  print(f"\nStopping training at batch {i} as requested")
                  break

              self.optimizer.zero_grad()

              # Wrap forward pass in autocast for mixed precision
              with torch.cuda.amp.autocast():
                  self._model.forward(
                      input_ids=data['input_ids'].to(self.device),
                      graph_pes=data['graph_pes'].to(self.device),
                      labels=data['labels'].to(self.device)
                  )

              # Scale the loss for backward pass
              # Note: We need to access the loss directly from the model
              if hasattr(self._model, 'loss_seg'):
                  scaler.scale(self._model.loss_seg).backward()
              else:
                  # Fallback to regular backward if loss not found
                  self._model.backward()

              # Unscale gradients before clipping
              scaler.unscale_(self.optimizer)

              # Clip gradients
              torch.nn.utils.clip_grad_norm_(self._model.parameters(), max_norm=1.0)

              # Optimizer step with scaler
              scaler.step(self.optimizer)
              scaler.update()

              # Scheduler step
              self.scheduler.step()

              # Track metrics every 10 batches
              if i % 10 == 0:
                  self.tracker.track(self._model)

              # Update progress bar every 50 batches
              if i % 50 == 0 and self.tracker.losses:
                  metrics = self.tracker.get_metrics()

                # Update progress bar with basic info
                  tq_train_loader.set_postfix(loss=f"{metrics['train_loss']:.4f}")

                  # Print detailed metrics
                  print(f"\n[Batch {i}/{len(self.train_loader)}]")
                  print(f"  Train Loss: {metrics['train_loss']:.4f}")
                  print(f"  Class 0 Acc: {metrics['train_class0_acc']:.2%} ({self.tracker.class_0_correct}/{self.tracker.class_0_total})")
                  print(f"  Class 1 Acc: {metrics['train_class1_acc']:.2%} ({self.tracker.class_1_correct}/{self.tracker.class_1_total})")

    def val_epoch(self):
        """Validation epoch"""
        self._model.eval()
        self.tracker.reset("val")

        with torch.no_grad():
            with tqdm(self.val_loader, desc="Validation") as tq_val_loader:
                for i, data in enumerate(tq_val_loader):
                    if self.max_batches and i >= self.max_batches // 5:
                        break

                    # Forward pass only
                    self._model.forward(
                        input_ids=data['input_ids'].to(self.device),
                        graph_pes=data['graph_pes'].to(self.device),
                        labels=data['labels'].to(self.device)
                    )

                    # Track metrics
                    self.tracker.track(self._model)

                    # Update progress bar
                    if self.tracker.losses:
                        tq_val_loader.set_postfix(**self.tracker.get_metrics())

    def evaluate(self, data_loader, split_name="test", max_batches=50):
        """Full evaluation on any data loader"""
        self._model.eval()
        self.tracker.reset(split_name)

        print(f"\nEvaluating on {split_name} set (first {max_batches} batches)...")

        with torch.no_grad():
            with tqdm(data_loader, desc=f"{split_name.capitalize()} Evaluation", total=min(max_batches, len(data_loader))) as tq_loader:
                for i, data in enumerate(tq_loader):
                    # Stop after max_batches
                    if i >= max_batches:
                        print(f"\nStopped evaluation at batch {i} as requested")
                        break

                    # Forward pass
                    self._model.forward(
                        input_ids=data['input_ids'].to(self.device),
                        graph_pes=data['graph_pes'].to(self.device),
                        labels=data['labels'].to(self.device)
                    )

                    # Track metrics
                    self.tracker.track(self._model)

                    # Update progress bar
                    if self.tracker.losses:
                        tq_loader.set_postfix(**self.tracker.get_metrics())

        # Get final metrics
        final_metrics = self.tracker.get_metrics()

        # Print detailed results
        print(f"\n{split_name.upper()} SET FINAL RESULTS:")
        print(f"  Loss: {final_metrics[f'{split_name}_loss']:.4f}")
        print(f"  Class 0 Acc: {final_metrics[f'{split_name}_class0_acc']:.2%}")
        print(f"  Class 1 Acc: {final_metrics[f'{split_name}_class1_acc']:.2%}")

        # Calculate additional metrics if we have both classes
        if self.tracker.class_1_total > 0:
            precision = self.tracker.class_1_correct / (self.tracker.class_1_correct + (self.tracker.class_0_total - self.tracker.class_0_correct)) if (self.tracker.class_1_correct + (self.tracker.class_0_total - self.tracker.class_0_correct)) > 0 else 0
            recall = final_metrics[f'{split_name}_class1_acc']
            f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0

            print(f"  Precision: {precision:.4f}")
            print(f"  Recall: {recall:.4f}")
            print(f"  F1 Score: {f1:.4f}")

        return {
            'loss': final_metrics[f'{split_name}_loss'],
            'class_0_acc': final_metrics[f'{split_name}_class0_acc'],
            'class_1_acc': final_metrics[f'{split_name}_class1_acc']
        }


# Usage example - now much cleaner!
if __name__ == "__main__":
    # Clear GPU memory before starting
    torch.cuda.empty_cache()
    gc.collect()

    # Setup device
    device = "cuda" if torch.cuda.is_available() else "cpu"
    print(f"Using device: {device}")

    # Training configuration
    BATCH_SIZE = 8
    EPOCHS = 1
    LEARNING_RATE = 1e-4
    MAX_BATCHES = 250

    # Create dataloaders
    print("\nCreating dataloaders...")
    train_loader, val_loader, test_loader, tokenizer = create_dataloaders(
        batch_size=BATCH_SIZE,
        tokenizer=None
    )

    # Store results for comparison
    results = {}

    # Train model WITH positional encodings
    print("\n" + "="*50)
    print("TRAINING MODEL WITH POSITIONAL ENCODINGS")
    print("="*50)

    model_with_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=True)
    trainer_with_pe = Trainer(
        model_with_pe,
        train_loader,
        val_loader,
        test_loader,  # Now passing test loader
        num_epochs=EPOCHS,
        learning_rate=LEARNING_RATE,
        device=device,
        max_batches=MAX_BATCHES
    )

    # fit() now returns test results!
    results['with_pe'] = trainer_with_pe.fit()

    # Save model
    #torch.save(model_with_pe.state_dict(), 'model_with_pe.pth')


    torch.cuda.empty_cache()
    gc.collect()

    # Train model WITHOUT positional encodings
    print("\n" + "="*50)
    print("TRAINING MODEL WITHOUT POSITIONAL ENCODINGS")
    print("="*50)

    model_without_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=False)
    trainer_without_pe = Trainer(
        model_without_pe,
        train_loader,
        val_loader,
        test_loader,  # Now passing test loader
        num_epochs=EPOCHS,
        learning_rate=LEARNING_RATE,
        device=device,
        max_batches=MAX_BATCHES
    )

    # fit() now returns test results!
    results['without_pe'] = trainer_without_pe.fit()

    # Save model
    #torch.save(model_without_pe.state_dict(), 'model_without_pe.pth')

    # Now comparison works because results is populated!
    print("\n" + "="*50)
    print("COMPARISON OF RESULTS")
    print("="*50)
    print(f"\nModel WITH Positional Encodings:")
    print(f"  Test Loss: {results['with_pe']['loss']:.4f}")
    print(f"  Class 0 Accuracy: {results['with_pe']['class_0_acc']:.2%}")
    print(f"  Class 1 Accuracy: {results['with_pe']['class_1_acc']:.2%}")

    print(f"\nModel WITHOUT Positional Encodings:")
    print(f"  Test Loss: {results['without_pe']['loss']:.4f}")
    print(f"  Class 0 Accuracy: {results['without_pe']['class_0_acc']:.2%}")
    print(f"  Class 1 Accuracy: {results['without_pe']['class_1_acc']:.2%}")

    print(f"\nDifferences (WITH PE - WITHOUT PE):")
    print(f"  Loss Difference: {results['with_pe']['loss'] - results['without_pe']['loss']:.4f}")
    print(f"  Class 0 Acc Difference: {results['with_pe']['class_0_acc'] - results['without_pe']['class_0_acc']:.2%}")
    print(f"  Class 1 Acc Difference: {results['with_pe']['class_1_acc'] - results['without_pe']['class_1_acc']:.2%}")

Using device: cuda

Creating dataloaders...

TRAINING MODEL WITH POSITIONAL ENCODINGS
Loading EleutherAI/pythia-160m...
Using positional encodings: True
Total parameters: 162,862,848
Trainable parameters: 539,904
Trainable %: 0.33%

=========== EPOCH 1 ===========


/tmp/ipython-input-4055256034.py:176: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


Training:   0%|          | 0/3084 [00:00<?, ?it/s]

/tmp/ipython-input-4055256034.py:187: FutureWarning: `torch.cuda.amp.autocast(args...)` is deprecated. Please use `torch.amp.autocast('cuda', args...)` instead.
  with torch.cuda.amp.autocast():



[Batch 0/3084]
  Train Loss: 0.2584
  Class 0 Acc: 0.00% (0/8)
  Class 1 Acc: 0.00% (0/0)

[Batch 50/3084]
  Train Loss: 0.1473
  Class 0 Acc: 65.96% (31/47)
  Class 1 Acc: 0.00% (0/1)

[Batch 100/3084]
  Train Loss: 0.0803
  Class 0 Acc: 81.18% (69/85)
  Class 1 Acc: 66.67% (2/3)

[Batch 150/3084]
  Train Loss: 0.0552
  Class 0 Acc: 86.99% (107/123)
  Class 1 Acc: 80.00% (4/5)

[Batch 200/3084]
  Train Loss: 0.0421
  Class 0 Acc: 90.06% (145/161)
  Class 1 Acc: 85.71% (6/7)

Stopping training at batch 250 as requested

TRAIN - Epoch 1
  Loss: 0.0353
  Class 0 Acc: 91.67%
  Class 1 Acc: 87.50%


Validation:   0%|          | 0/1028 [00:00<?, ?it/s]


VAL - Epoch 1
  Loss: 0.0000
  Class 0 Acc: 100.00%
  Class 1 Acc: 100.00%

FINAL TEST SET EVALUATION

Evaluating on test set (first 50 batches)...


Test Evaluation:   0%|          | 0/50 [00:00<?, ?it/s]


Stopped evaluation at batch 50 as requested

TEST SET FINAL RESULTS:
  Loss: 0.0000
  Class 0 Acc: 100.00%
  Class 1 Acc: 100.00%
  Precision: 1.0000
  Recall: 1.0000
  F1 Score: 1.0000

TRAINING MODEL WITHOUT POSITIONAL ENCODINGS
Loading EleutherAI/pythia-160m...
Using positional encodings: False
Total parameters: 162,839,040
Trainable parameters: 516,096
Trainable %: 0.32%

=========== EPOCH 1 ===========


Training:   0%|          | 0/3084 [00:00<?, ?it/s]


[Batch 0/3084]
  Train Loss: 1.0662
  Class 0 Acc: 0.00% (0/7)
  Class 1 Acc: 0.00% (0/1)

[Batch 50/3084]
  Train Loss: 0.2167
  Class 0 Acc: 63.83% (30/47)
  Class 1 Acc: 0.00% (0/1)

[Batch 100/3084]
  Train Loss: 0.1182
  Class 0 Acc: 80.23% (69/86)
  Class 1 Acc: 50.00% (1/2)

[Batch 150/3084]
  Train Loss: 0.0814
  Class 0 Acc: 86.29% (107/124)
  Class 1 Acc: 75.00% (3/4)

[Batch 200/3084]
  Train Loss: 0.0620
  Class 0 Acc: 89.44% (144/161)
  Class 1 Acc: 85.71% (6/7)

Stopping training at batch 250 as requested

TRAIN - Epoch 1
  Loss: 0.0521
  Class 0 Acc: 91.19%
  Class 1 Acc: 85.71%


Validation:   0%|          | 0/1028 [00:00<?, ?it/s]


VAL - Epoch 1
  Loss: 0.0000
  Class 0 Acc: 100.00%
  Class 1 Acc: 100.00%

FINAL TEST SET EVALUATION

Evaluating on test set (first 50 batches)...


Test Evaluation:   0%|          | 0/50 [00:00<?, ?it/s]


Stopped evaluation at batch 50 as requested

TEST SET FINAL RESULTS:
  Loss: 0.0000
  Class 0 Acc: 100.00%
  Class 1 Acc: 100.00%
  Precision: 1.0000
  Recall: 1.0000
  F1 Score: 1.0000

COMPARISON OF RESULTS

Model WITH Positional Encodings:
  Test Loss: 0.0000
  Class 0 Accuracy: 100.00%
  Class 1 Accuracy: 100.00%

Model WITHOUT Positional Encodings:
  Test Loss: 0.0000
  Class 0 Accuracy: 100.00%
  Class 1 Accuracy: 100.00%

Differences (WITH PE - WITHOUT PE):
  Loss Difference: -0.0000
  Class 0 Acc Difference: 0.00%
  Class 1 Acc Difference: 0.00%


In [28]:
#We analyze the activations to check for dead neurons and abnormal values, we notice high
def analyze_activations(model, data_loader):
    activations = []

    def hook_fn(module, input, output):
        activations.append(output.detach())

    # Register hooks on key layers
    hooks = []
    for name, module in model.named_modules():
        if isinstance(module, nn.Linear):
            hooks.append(module.register_forward_hook(hook_fn))

    # Get device from model
    device = next(model.parameters()).device

    # Run one batch
    batch = next(iter(data_loader))
    with torch.no_grad():
        # Move data to same device as model
        input_ids = batch['input_ids'].to(device)
        graph_pes = batch['graph_pes'].to(device)
        model(input_ids, graph_pes)

    # Analyze
    for i, act in enumerate(activations):
        print(f"Layer {i}: mean={act.mean():.3f}, std={act.std():.3f}, "
              f"% dead neurons={(act == 0).float().mean():.3f}")

    # Clean up hooks
    for hook in hooks:
        hook.remove()

# Now run it
analyze_activations(model_without_pe, test_loader)



Layer 0: mean=0.014, std=1.773, % dead neurons=0.000
Layer 1: mean=0.615, std=3.374, % dead neurons=0.000
Layer 2: mean=0.000, std=0.004, % dead neurons=0.000
Layer 3: mean=-0.002, std=0.251, % dead neurons=0.000
Layer 4: mean=-0.406, std=0.626, % dead neurons=0.000
Layer 5: mean=1.347, std=4.883, % dead neurons=0.000
Layer 6: mean=-0.000, std=0.007, % dead neurons=0.000
Layer 7: mean=-0.008, std=0.564, % dead neurons=0.000
Layer 8: mean=0.598, std=3.350, % dead neurons=0.000
Layer 9: mean=-0.000, std=0.004, % dead neurons=0.000
Layer 10: mean=-0.038, std=1.861, % dead neurons=0.000
Layer 11: mean=-0.835, std=7.491, % dead neurons=0.000
Layer 12: mean=0.000, std=0.012, % dead neurons=0.000
Layer 13: mean=0.001, std=0.174, % dead neurons=0.000
Layer 14: mean=-0.421, std=1.073, % dead neurons=0.000
Layer 15: mean=-2.506, std=7.839, % dead neurons=0.000
Layer 16: mean=-0.000, std=0.012, % dead neurons=0.000
Layer 17: mean=0.004, std=1.406, % dead neurons=0.000
Layer 18: mean=1.527, std=7.